File: scenario_mpc/scenario_pruning.py
"""
Scenario pruning for efficient optimization.

Implements Algorithms 3 and 4 from guide.md:
- Algorithm 3: Geometric dominance pruning
- Algorithm 4: Support-based scenario removal

These algorithms reduce the number of scenarios while maintaining
chance constraint guarantees.
"""

import numpy as np
from typing import List, Set, Tuple, Optional
from .types import (
    EgoState,
    Scenario,
    CollisionConstraint,
)


def prune_dominated_scenarios(
    scenarios: List[Scenario],
    reference_trajectory: List[EgoState],
    ego_radius: float,
    obstacle_radius: float
) -> List[Scenario]:
    """
    Prune geometrically dominated scenarios.

    Algorithm 3 from guide.md: GeometricDominancePruning

    A scenario s1 dominates s2 if for all timesteps k and obstacles o,
    the obstacle position in s1 is further from the ego than in s2.

    Args:
        scenarios: List of scenarios to prune
        reference_trajectory: Reference ego trajectory
        ego_radius: Ego collision radius
        obstacle_radius: Obstacle collision radius

    Returns:
        Pruned list of non-dominated scenarios
    """
    if len(scenarios) <= 1:
        return scenarios

    n = len(scenarios)
    dominated = set()

    # Pairwise comparison
    for i in range(n):
        if i in dominated:
            continue

        for j in range(i + 1, n):
            if j in dominated:
                continue

            # Check if i dominates j or vice versa
            dom_i_j = _scenario_dominates(
                scenarios[i], scenarios[j], reference_trajectory
            )
            dom_j_i = _scenario_dominates(
                scenarios[j], scenarios[i], reference_trajectory
            )

            if dom_i_j:
                dominated.add(j)
            elif dom_j_i:
                dominated.add(i)
                break

    # Return non-dominated scenarios
    return [s for i, s in enumerate(scenarios) if i not in dominated]


def _scenario_dominates(
    s1: Scenario,
    s2: Scenario,
    reference_trajectory: List[EgoState]
) -> bool:
    """
    Check if scenario s1 dominates s2.

    s1 dominates s2 if obstacles in s1 are always at least as far
    from the ego as in s2 (s1 is "safer" everywhere).
    """
    horizon = len(reference_trajectory) - 1

    for k in range(horizon + 1):
        ego_pos = reference_trajectory[k].position()

        # Check all obstacles present in both scenarios
        common_obs = set(s1.trajectories.keys()) & set(s2.trajectories.keys())

        for obs_id in common_obs:
            traj1 = s1.trajectories[obs_id]
            traj2 = s2.trajectories[obs_id]

            if k >= len(traj1.steps) or k >= len(traj2.steps):
                continue

            pos1 = traj1.steps[k].mean
            pos2 = traj2.steps[k].mean

            dist1 = np.linalg.norm(ego_pos - pos1)
            dist2 = np.linalg.norm(ego_pos - pos2)

            # If s2 obstacle is further at any point, s1 doesn't dominate
            if dist2 > dist1 + 1e-6:
                return False

    return True


def remove_inactive_scenarios(
    scenarios: List[Scenario],
    constraints: List[CollisionConstraint],
    optimal_trajectory: List[EgoState],
    tolerance: float = 1e-4
) -> Tuple[List[Scenario], Set[int]]:
    """
    Remove scenarios with inactive constraints.

    Algorithm 4 from guide.md: SupportBasedRemoval

    After solving the optimization, identify scenarios whose constraints
    are not active (not binding) and remove them.

    A constraint is active if it is satisfied with equality (within tolerance).

    Args:
        scenarios: Current list of scenarios
        constraints: Collision constraints for all scenarios
        optimal_trajectory: Optimal ego trajectory from solver
        tolerance: Tolerance for checking constraint activity

    Returns:
        (remaining_scenarios, active_scenario_ids)
    """
    # Group constraints by scenario
    constraints_by_scenario = {}
    for c in constraints:
        if c.scenario_id not in constraints_by_scenario:
            constraints_by_scenario[c.scenario_id] = []
        constraints_by_scenario[c.scenario_id].append(c)

    active_scenarios = set()

    for scenario in scenarios:
        sid = scenario.scenario_id

        if sid not in constraints_by_scenario:
            continue

        # Check if any constraint from this scenario is active
        for constraint in constraints_by_scenario[sid]:
            k = constraint.k
            if k >= len(optimal_trajectory):
                continue

            ego_pos = optimal_trajectory[k].position()
            value = constraint.evaluate(ego_pos)

            # Constraint is active if value is close to 0
            if abs(value) < tolerance:
                active_scenarios.add(sid)
                break

    # Keep scenarios that have active constraints
    remaining = [s for s in scenarios if s.scenario_id in active_scenarios]

    # If no scenarios are active, keep all (safety fallback)
    if not remaining:
        return scenarios, set(s.scenario_id for s in scenarios)

    return remaining, active_scenarios


def prune_by_probability(
    scenarios: List[Scenario],
    min_probability: float = 0.01
) -> List[Scenario]:
    """
    Remove scenarios with very low probability.

    Simple pruning that removes scenarios unlikely to occur.

    Args:
        scenarios: List of scenarios
        min_probability: Minimum probability threshold

    Returns:
        Filtered scenarios
    """
    return [s for s in scenarios if s.probability >= min_probability]


def cluster_similar_scenarios(
    scenarios: List[Scenario],
    distance_threshold: float = 1.0,
    horizon: int = None
) -> List[Scenario]:
    """
    Cluster similar scenarios and keep representative ones.

    Scenarios are similar if their obstacle trajectories are close
    across the horizon.

    Args:
        scenarios: List of scenarios
        distance_threshold: Maximum distance for clustering
        horizon: Prediction horizon (inferred if not provided)

    Returns:
        Representative scenarios (cluster centroids)
    """
    if len(scenarios) <= 1:
        return scenarios

    # Compute scenario features (flattened obstacle positions)
    features = []
    for scenario in scenarios:
        feat = _scenario_to_feature(scenario, horizon)
        features.append(feat)

    features = np.array(features)

    # Simple greedy clustering
    kept_indices = []
    for i, feat in enumerate(features):
        is_close = False
        for j in kept_indices:
            if np.linalg.norm(feat - features[j]) < distance_threshold:
                is_close = True
                break
        if not is_close:
            kept_indices.append(i)

    return [scenarios[i] for i in kept_indices]


def _scenario_to_feature(scenario: Scenario, horizon: int = None) -> np.ndarray:
    """
    Convert scenario to feature vector for clustering.
    """
    if horizon is None:
        # Infer horizon from trajectory length
        for traj in scenario.trajectories.values():
            horizon = len(traj.steps) - 1
            break
        if horizon is None:
            horizon = 10

    features = []

    # Sort obstacle IDs for consistent ordering
    for obs_id in sorted(scenario.trajectories.keys()):
        traj = scenario.trajectories[obs_id]
        for step in traj.steps[:horizon + 1]:
            features.extend(step.mean)

    return np.array(features) if features else np.array([0.0])


def select_diverse_scenarios(
    scenarios: List[Scenario],
    num_select: int,
    horizon: int = None
) -> List[Scenario]:
    """
    Select a diverse subset of scenarios.

    Uses farthest point sampling to maximize diversity.

    Args:
        scenarios: All scenarios
        num_select: Number of scenarios to select
        horizon: Prediction horizon

    Returns:
        Diverse subset of scenarios
    """
    if len(scenarios) <= num_select:
        return scenarios

    # Convert to features
    features = np.array([_scenario_to_feature(s, horizon) for s in scenarios])

    # Farthest point sampling
    selected = [0]  # Start with first scenario

    while len(selected) < num_select:
        # Find point farthest from all selected points
        max_min_dist = -1
        best_idx = -1

        for i in range(len(scenarios)):
            if i in selected:
                continue

            # Min distance to any selected point
            min_dist = float('inf')
            for j in selected:
                dist = np.linalg.norm(features[i] - features[j])
                min_dist = min(min_dist, dist)

            if min_dist > max_min_dist:
                max_min_dist = min_dist
                best_idx = i

        if best_idx >= 0:
            selected.append(best_idx)
        else:
            break

    return [scenarios[i] for i in selected]


def adaptive_scenario_budget(
    current_scenarios: int,
    constraint_violations: float,
    max_scenarios: int = 100,
    min_scenarios: int = 5,
    growth_factor: float = 1.5,
    shrink_factor: float = 0.8
) -> int:
    """
    Adaptively adjust scenario budget based on constraint violations.

    If constraints are frequently violated, increase scenarios.
    If constraints are easily satisfied, decrease scenarios.

    Args:
        current_scenarios: Current number of scenarios
        constraint_violations: Sum of constraint violations
        max_scenarios: Maximum allowed scenarios
        min_scenarios: Minimum scenarios to maintain
        growth_factor: Factor to increase scenarios
        shrink_factor: Factor to decrease scenarios

    Returns:
        New scenario budget
    """
    if constraint_violations > 0.1:
        # Violations present - increase scenarios
        new_budget = int(current_scenarios * growth_factor)
    elif constraint_violations < 0.001:
        # No violations - can decrease scenarios
        new_budget = int(current_scenarios * shrink_factor)
    else:
        new_budget = current_scenarios

    return max(min_scenarios, min(max_scenarios, new_budget))

File: cpp_mpc/src/scenario_pruning.cpp
/**
 * @file scenario_pruning.cpp
 * @brief Implementation of scenario pruning algorithms.
 */

#include "scenario_pruning.hpp"
#include <algorithm>
#include <cmath>
#include <limits>

namespace scenario_mpc {

namespace {

/**
 * @brief Check if scenario s1 dominates s2.
 *
 * s1 dominates s2 if obstacles in s1 are always at least as CLOSE to
 * the ego as in s2 (s1 is more constraining everywhere).
 * If s1 dominates s2, then s2 can be pruned because satisfying s1's
 * constraints automatically satisfies s2's constraints.
 */
bool scenario_dominates(
    const Scenario& s1,
    const Scenario& s2,
    const std::vector<EgoState>& reference_trajectory
) {
    int horizon = static_cast<int>(reference_trajectory.size()) - 1;

    for (int k = 0; k <= horizon; ++k) {
        Eigen::Vector2d ego_pos = reference_trajectory[k].position();

        // Check all obstacles present in both scenarios
        for (const auto& [obs_id, traj1] : s1.trajectories) {
            auto it2 = s2.trajectories.find(obs_id);
            if (it2 == s2.trajectories.end()) {
                continue;
            }
            const ObstacleTrajectory& traj2 = it2->second;

            if (k >= static_cast<int>(traj1.steps.size()) ||
                k >= static_cast<int>(traj2.steps.size())) {
                continue;
            }

            Eigen::Vector2d pos1 = traj1.steps[k].mean;
            Eigen::Vector2d pos2 = traj2.steps[k].mean;

            double dist1 = (ego_pos - pos1).norm();
            double dist2 = (ego_pos - pos2).norm();

            // s1 dominates s2 if s1's obstacles are closer (more constraining)
            // If s1's obstacle is farther at any point, s1 doesn't dominate
            if (dist1 > dist2 + 1e-6) {
                return false;
            }
        }
    }

    return true;
}

/**
 * @brief Convert scenario to low-dimensional quotient feature (mean/end positions per obstacle).
 * Used for quotient-space reduction to cluster in reduced dimension.
 */
Eigen::VectorXd scenario_to_quotient_feature(const Scenario& scenario, int horizon = -1) {
    if (scenario.trajectories.empty()) {
        return Eigen::VectorXd::Zero(1);
    }
    std::vector<int> sorted_obs;
    for (const auto& [obs_id, _] : scenario.trajectories) {
        sorted_obs.push_back(obs_id);
    }
    std::sort(sorted_obs.begin(), sorted_obs.end());
    std::vector<double> feats;
    for (int obs_id : sorted_obs) {
        const auto& traj = scenario.trajectories.at(obs_id);
        if (traj.steps.empty()) continue;
        int H = horizon >= 0 ? std::min(horizon + 1, static_cast<int>(traj.steps.size())) : static_cast<int>(traj.steps.size());
        double sum_x = 0, sum_y = 0;
        for (int k = 0; k < H; ++k) {
            sum_x += traj.steps[k].mean(0);
            sum_y += traj.steps[k].mean(1);
        }
        double mean_x = sum_x / H;
        double mean_y = sum_y / H;
        double end_x = traj.steps[H - 1].mean(0);
        double end_y = traj.steps[H - 1].mean(1);
        feats.push_back(mean_x);
        feats.push_back(mean_y);
        feats.push_back(end_x);
        feats.push_back(end_y);
    }
    if (feats.empty()) return Eigen::VectorXd::Constant(1, 0.0);
    Eigen::VectorXd out(feats.size());
    for (size_t i = 0; i < feats.size(); ++i) out(i) = feats[i];
    return out;
}

/**
 * @brief Convert scenario to feature vector for clustering.
 */
Eigen::VectorXd scenario_to_feature(const Scenario& scenario, int horizon = -1) {
    if (horizon < 0) {
        // Infer horizon from trajectory length
        for (const auto& [_, traj] : scenario.trajectories) {
            horizon = static_cast<int>(traj.steps.size()) - 1;
            break;
        }
        if (horizon < 0) {
            horizon = 10;
        }
    }

    std::vector<double> features;

    // Sort obstacle IDs for consistent ordering
    std::vector<int> sorted_obs_ids;
    for (const auto& [obs_id, _] : scenario.trajectories) {
        sorted_obs_ids.push_back(obs_id);
    }
    std::sort(sorted_obs_ids.begin(), sorted_obs_ids.end());

    for (int obs_id : sorted_obs_ids) {
        const auto& traj = scenario.trajectories.at(obs_id);
        int num_steps = std::min(horizon + 1, static_cast<int>(traj.steps.size()));
        for (int k = 0; k < num_steps; ++k) {
            features.push_back(traj.steps[k].mean(0));
            features.push_back(traj.steps[k].mean(1));
        }
    }

    if (features.empty()) {
        return Eigen::VectorXd::Constant(1, 0.0);
    }

    Eigen::VectorXd result(features.size());
    for (size_t i = 0; i < features.size(); ++i) {
        result(i) = features[i];
    }
    return result;
}

}  // anonymous namespace

std::vector<Scenario> prune_dominated_scenarios(
    const std::vector<Scenario>& scenarios,
    const std::vector<EgoState>& reference_trajectory,
    double ego_radius,
    double obstacle_radius
) {
    if (scenarios.size() <= 1) {
        return scenarios;
    }

    int n = static_cast<int>(scenarios.size());
    std::set<int> dominated;

    // Pairwise comparison (never prune DRO-injected scenarios)
    for (int i = 0; i < n; ++i) {
        if (dominated.count(i) || scenarios[i].is_injected) {
            continue;
        }

        for (int j = i + 1; j < n; ++j) {
            if (dominated.count(j)) {
                continue;
            }

            // Never prune an injected scenario
            if (scenarios[j].is_injected) {
                continue;
            }

            // Check if i dominates j or vice versa
            bool dom_i_j = scenario_dominates(
                scenarios[i], scenarios[j], reference_trajectory
            );
            bool dom_j_i = scenario_dominates(
                scenarios[j], scenarios[i], reference_trajectory
            );

            if (dom_i_j) {
                dominated.insert(j);
            } else if (dom_j_i) {
                dominated.insert(i);
                break;
            }
        }
    }

    // Return non-dominated scenarios
    std::vector<Scenario> result;
    for (int i = 0; i < n; ++i) {
        if (dominated.count(i) == 0) {
            result.push_back(scenarios[i]);
        }
    }
    return result;
}

std::pair<std::vector<Scenario>, std::set<int>> remove_inactive_scenarios(
    const std::vector<Scenario>& scenarios,
    const std::vector<CollisionConstraint>& constraints,
    const std::vector<EgoState>& optimal_trajectory,
    double tolerance
) {
    // Group constraints by scenario
    std::map<int, std::vector<CollisionConstraint>> constraints_by_scenario;
    for (const auto& c : constraints) {
        constraints_by_scenario[c.scenario_id].push_back(c);
    }

    std::set<int> active_scenarios;

    for (const auto& scenario : scenarios) {
        int sid = scenario.scenario_id;

        if (constraints_by_scenario.find(sid) == constraints_by_scenario.end()) {
            continue;
        }

        // Check if any constraint from this scenario is active
        for (const auto& constraint : constraints_by_scenario[sid]) {
            int k = constraint.k;
            if (k >= static_cast<int>(optimal_trajectory.size())) {
                continue;
            }

            Eigen::Vector2d ego_pos = optimal_trajectory[k].position();
            double value = constraint.evaluate(ego_pos);

            // Constraint is active if value is close to 0
            if (std::abs(value) < tolerance) {
                active_scenarios.insert(sid);
                break;
            }
        }
    }

    // Keep scenarios that have active constraints
    std::vector<Scenario> remaining;
    for (const auto& s : scenarios) {
        if (active_scenarios.count(s.scenario_id)) {
            remaining.push_back(s);
        }
    }

    // If no scenarios are active, keep all (safety fallback)
    if (remaining.empty()) {
        std::set<int> all_ids;
        for (const auto& s : scenarios) {
            all_ids.insert(s.scenario_id);
        }
        return {scenarios, all_ids};
    }

    return {remaining, active_scenarios};
}

std::vector<Scenario> prune_by_probability(
    const std::vector<Scenario>& scenarios,
    double min_probability
) {
    std::vector<Scenario> filtered;
    for (const auto& s : scenarios) {
        if (s.probability >= min_probability) {
            filtered.push_back(s);
        }
    }
    return filtered;
}

std::vector<Scenario> cluster_similar_scenarios(
    const std::vector<Scenario>& scenarios,
    double distance_threshold,
    int horizon
) {
    if (scenarios.size() <= 1) {
        return scenarios;
    }

    // Compute scenario features (flattened obstacle positions)
    std::vector<Eigen::VectorXd> features;
    for (const auto& scenario : scenarios) {
        features.push_back(scenario_to_feature(scenario, horizon));
    }

    // Simple greedy clustering
    std::vector<int> kept_indices;
    for (size_t i = 0; i < features.size(); ++i) {
        bool is_close = false;
        for (int j : kept_indices) {
            if ((features[i] - features[j]).norm() < distance_threshold) {
                is_close = true;
                break;
            }
        }
        if (!is_close) {
            kept_indices.push_back(static_cast<int>(i));
        }
    }

    std::vector<Scenario> result;
    for (int i : kept_indices) {
        result.push_back(scenarios[i]);
    }
    return result;
}

std::vector<Scenario> select_diverse_scenarios(
    const std::vector<Scenario>& scenarios,
    int num_select,
    int horizon
) {
    if (static_cast<int>(scenarios.size()) <= num_select) {
        return scenarios;
    }

    // Convert to features
    std::vector<Eigen::VectorXd> features;
    for (const auto& scenario : scenarios) {
        features.push_back(scenario_to_feature(scenario, horizon));
    }

    // Farthest point sampling
    std::vector<int> selected;
    selected.push_back(0);  // Start with first scenario

    while (static_cast<int>(selected.size()) < num_select) {
        // Find point farthest from all selected points
        double max_min_dist = -1;
        int best_idx = -1;

        for (int i = 0; i < static_cast<int>(scenarios.size()); ++i) {
            if (std::find(selected.begin(), selected.end(), i) != selected.end()) {
                continue;
            }

            // Min distance to any selected point
            double min_dist = std::numeric_limits<double>::infinity();
            for (int j : selected) {
                double dist = (features[i] - features[j]).norm();
                min_dist = std::min(min_dist, dist);
            }

            if (min_dist > max_min_dist) {
                max_min_dist = min_dist;
                best_idx = i;
            }
        }

        if (best_idx >= 0) {
            selected.push_back(best_idx);
        } else {
            break;
        }
    }

    std::vector<Scenario> result;
    for (int i : selected) {
        result.push_back(scenarios[i]);
    }
    return result;
}

int adaptive_scenario_budget(
    int current_scenarios,
    double constraint_violations,
    int max_scenarios,
    int min_scenarios,
    double growth_factor,
    double shrink_factor
) {
    int new_budget;

    if (constraint_violations > 0.1) {
        // Violations present - increase scenarios
        new_budget = static_cast<int>(current_scenarios * growth_factor);
    } else if (constraint_violations < 0.001) {
        // No violations - can decrease scenarios
        new_budget = static_cast<int>(current_scenarios * shrink_factor);
    } else {
        new_budget = current_scenarios;
    }

    return std::max(min_scenarios, std::min(max_scenarios, new_budget));
}

std::vector<Scenario> reduce_scenarios_quotient_space(
    const std::vector<Scenario>& scenarios,
    int num_quotient,
    int horizon
) {
    const int S = static_cast<int>(scenarios.size());
    if (S <= num_quotient || num_quotient <= 0) {
        return scenarios;
    }
    int H = horizon;
    if (H < 0 && !scenarios.empty() && !scenarios[0].trajectories.empty()) {
        H = static_cast<int>(scenarios[0].trajectories.begin()->second.steps.size()) - 1;
    }
    if (H < 0) H = 10;

    std::vector<Eigen::VectorXd> qfeatures;
    qfeatures.reserve(S);
    for (const auto& sc : scenarios) {
        qfeatures.push_back(scenario_to_quotient_feature(sc, H));
    }
    const int dim = static_cast<int>(qfeatures[0].size());

    // K-centers: iteratively add the point farthest from current centers
    std::vector<int> centers;
    centers.push_back(0);
    while (static_cast<int>(centers.size()) < num_quotient) {
        double max_dist = -1;
        int best_idx = -1;
        for (int i = 0; i < S; ++i) {
            double min_d = std::numeric_limits<double>::infinity();
            for (int c : centers) {
                double d = (qfeatures[i] - qfeatures[c]).norm();
                min_d = std::min(min_d, d);
            }
            if (min_d > max_dist) {
                max_dist = min_d;
                best_idx = i;
            }
        }
        if (best_idx >= 0) {
            centers.push_back(best_idx);
        } else {
            break;
        }
    }

    // Assign each scenario to nearest center; compute cluster centroid
    int K = static_cast<int>(centers.size());
    std::vector<Eigen::VectorXd> centroids(K);
    std::vector<std::vector<int>> clusters(K);
    for (int i = 0; i < K; ++i) {
        centroids[i] = Eigen::VectorXd::Zero(dim);
    }
    for (int i = 0; i < S; ++i) {
        int nearest = 0;
        double d_min = (qfeatures[i] - qfeatures[centers[0]]).norm();
        for (int k = 1; k < K; ++k) {
            double d = (qfeatures[i] - qfeatures[centers[k]]).norm();
            if (d < d_min) {
                d_min = d;
                nearest = k;
            }
        }
        clusters[nearest].push_back(i);
        centroids[nearest] += qfeatures[i];
    }
    for (int k = 0; k < K; ++k) {
        if (!clusters[k].empty()) {
            centroids[k] /= static_cast<double>(clusters[k].size());
        }
    }

    // Representative per cluster: scenario closest to cluster centroid
    std::vector<Scenario> result;
    result.reserve(K);
    for (int k = 0; k < K; ++k) {
        if (clusters[k].empty()) {
            result.push_back(scenarios[centers[k]]);
            continue;
        }
        int rep = clusters[k][0];
        double rep_dist = (qfeatures[rep] - centroids[k]).norm();
        for (int idx : clusters[k]) {
            double d = (qfeatures[idx] - centroids[k]).norm();
            if (d < rep_dist) {
                rep_dist = d;
                rep = idx;
            }
        }
        Scenario sc = scenarios[rep];
        sc.scenario_id = k;
        result.push_back(sc);
    }
    return result;
}

}  // namespace scenario_mpc
